Provider component for Commerce API client

vmarta opened the pull request:
This PR implements the Provider component for Commerce API client:
- The Provider initializes the api client. And it will be re-initialized whenever a different prop to the Provider is passed in.
- There's a `useCommerceApi` hook available to consume the context provider

To show that things are semi working, you can run the typescript-minimal template.

Todos:
- [x] Document the provider and hook
- [x] Write tests
- [x] Revert changes to typescript-minimal template

# How to Test-Drive This PR

1. `npm ci` at the root
2. Switch to the typescript-minimal template directory
3. `npm run start`
4. Go to your browser console


# Checklists



## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

kevinxh committed:
add ShopperBasket stubs

kevinxh committed:
add comments

kevinxh committed:
update comment

kevinxh committed:
add QueryParams

kevinxh committed:
add ShopperProducts

kevinxh committed:
add ShopperSearch

kevinxh committed:
add ShopperPromotion

kevinxh committed:
add ShopperOrders

kevinxh committed:
add ShopperCustomer

kevinxh committed:
use optional property types

kevinxh committed:
add copyright header

kevinxh committed:
rename useShopperSearchProductSearch

kevinxh committed:
add type DependencyList

vmarta committed:
Create context for the api client

kevinxh committed:
use convention T

kevinxh committed:
add copyright header

kevinxh committed:
export types inline

kevinxh committed:
add children type

kevinxh committed:
update hooks type path

kevinxh committed:
add index file and consolidate exports

vmarta committed:
Copy api client code from retail-react-app

kevinxh committed:
define overloads for basket action hooks

vmarta committed:
Initialize api client

vmarta committed:
Rough porting the api-client code from retail-react-app

- mock some function and class, so that we're not importing pwa-kit-react-sdk
- add npm dependencies for the necessary 3rd party libs

vmarta committed:
Trying out the commerce api provider in this template

kevinxh committed:
fix types

kevinxh committed:
update implementation priority

vmarta committed:
Add comment

vmarta committed:
Merge branch 'commerce-sdk-react-stubs' into commerce-api-provider

kevinxh committed:
update shopper order priority

kevinxh committed:
Merge branch 'develop' into commerce-sdk-react-stubs

vmarta committed:
Delete unnecessary hooks

vmarta committed:
Delete redundant `children`

vmarta committed:
Merge branch 'commerce-sdk-react-stubs' into commerce-api-provider

vmarta committed:
Update package-lock.json

vmarta committed:
Some code cleanup

kevinxh committed:
auto lint fix

vmarta committed:
Tweak exports

vmarta committed:
Rename variables

kevinxh committed:
add ShopperLogin folder

kevinxh committed:
add useShopperLoginHelper

vmarta committed:
Remove unnecessary file

vmarta committed:
Merge branch 'commerce-sdk-react-stubs' into commerce-api-provider

vmarta committed:
Sync with parent branch

vmarta committed:
Logging the token

vmarta commented on packages/commerce-sdk-react/src/api-client/__mocks__/einstein.js:
```diff
@@ -0,0 +1,62 @@
+/*
```
Everything in the `api-client` folder is essentially copied from `template-retail-react-app/app/commerce-api/`

vmarta commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -17,8 +20,33 @@ export interface CommerceAPIProviderProps {
+    }
+
+    // TODO: auth logic should use the helpers from commerce-sdk-isomorphic ?
+    const commerceAPI = new CommerceAPI(config)
```
This `CommerceAPI` is the CommerceAPI class taken from the retail-react-app. The class does a few things, including:
- using `Proxy` to automatically inject the access token to the headers of API requests
- does manage the tokens and save them to either cookie or local storage.

vmarta replied:
@kevinxh I saw the comment about `useShopperLoginHelpers` managing the tokens. How about if we delegate it to a class like this instead?

vmarta commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -17,8 +20,33 @@ export interface CommerceAPIProviderProps {
     currency: string
 }
 
+export const CommerceAPIContext = React.createContext({} as ApiClients)
```
TIL `createContext` requires a default value to be passed in. So in Typescript, we can pass in an empty object and casting the type with `as` like that.

vmarta commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -17,8 +20,33 @@ export interface CommerceAPIProviderProps {
+    // TODO: use Context from useServerEffect
+    // See Kevin's PR: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/654/files#r914097886
+    // See Ben's PR: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/642
+    return <CommerceAPIContext.Provider value={commerceAPI}>{children}</CommerceAPIContext.Provider>
```
Ticket's acceptance criteria ask for integrating the `useServerEffect`. But since that work is still in progress and done in parallel, I can't integrate it here right now.

kevinxh commented on packages/commerce-sdk-react/src/api-client/pkce.js:
```diff
@@ -0,0 +1,54 @@
+/*
```
we don't need any PKCE code as that's implemented in the commerce-sdk-isomorpic library. 

(In retail react app, we haven't refactor to use the SLAS helpers but we should do that in our hook library)

kevinxh replied:
see https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/blob/main/src/static/helpers/slasHelper.ts

kevinxh commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -17,8 +20,33 @@ export interface CommerceAPIProviderProps {
+        throwOnBadResponse: true,
+    }
+
+    // TODO: auth logic should use the helpers from commerce-sdk-isomorphic ?
```
YEP! 👍

vmarta committed:
No longer exposing CommerceAPI class

As a result, the access token is not accessible directly by the end users. They would need to call one of the hooks to access it.

vmarta committed:
Share the auth object that will manage the tokens

kevinxh committed:
upgrade isomorphic lib to 1.7.0

vmarta committed:
Linting Typescript code with typescript-eslint plugin

vmarta committed:
Convert another rule to a warning

kevinxh committed:
Merge branch 'typescript-eslint-2' into commerce-sdk-react-stubs

kevinxh committed:
run npm run format

vmarta committed:
Merge branch 'commerce-sdk-react-stubs' into commerce-api-provider

vmarta committed:
Update package-lock.json

kevinxh committed:
install node types

kevinxh committed:
fix letter casing

kevinxh committed:
merge develop

kevinxh committed:
fix lint

vmarta committed:
Merge branch 'commerce-sdk-react-stubs' into commerce-api-provider

vmarta committed:
Simplify the Provider implementation

- No more copying over the api-client logic from retail-react-app
- Now the PR is simply a context provider and consumer.

vmarta committed:
Use Kevin's sample code for the api clients

vmarta committed:
More consistent casing

wjhsf commented on packages/commerce-sdk-react/package.json:
```diff
@@ -1,56 +1,57 @@
+  "types": "dist/index.d.ts",
+  "private": true,
+  "devDependencies": {
+    "@types/node": "^18.0.6",
```
This should be v14, since that's the oldest version of node that we support.

vmarta replied:
Good catch

wjhsf commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -0,0 +1,74 @@
+    currency: string
+}
+
+export const CommerceApiContext = React.createContext({} as ApiClients)
```
Type assertions (`thing as Type`) are essentially saying to TypeScript, "I know more than you do. Trust me." If we accessed things _immediately_, like `(null as Record<string, unknown>).property`, then TypeScript would allow it, and we'd end up a runtime error. On the flip side, as a contrived example, sometimes we need to build objects up as we go:
```typescript
const thing = {} as { foo: string }
if (condition) thing.foo = 'foo'
else thing.foo = 'not foo'
const useThing = (arg: { foo: string }) => {}
useThing(thing)
```
In this (contrived) case, the type assertion is fine because we immediately conform to the type we asserted.

Anyway, I don't know how React contexts work. Can we guarantee that the context will be replaced with an actual `ApiClients` object before anyone uses it? If not, then the type assertion isn't strictly safe.

Also, since type assertions are essentially lying to TypeScript, a good practice to follow is comments explaining why the assertion is necessary and why it is safe (or not) to do so, particularly any assumptions about the state of the system.

vmarta replied:
This is the case of "build objects up as we go". `React.createContext` requires an initial value to be passed in. In our case, it will be the set of api clients. But we can't initialize them until later.

> Can we guarantee that the context will be replaced with an actual ApiClients object before anyone uses it?

Yes, the context will have the actual ApiClients object:
- The Context Provider will be mounted first and will initialize the api clients
- Then the children component will be mounted. So if it calls `useCommerceApi` to consume the context, the api clients will already be initialized.

wjhsf commented on packages/commerce-sdk-react/src/hooks/types.ts:
```diff
@@ -0,0 +1,93 @@
+    ShopperSearchTypes.ShopperSearchParameters & Record<string, unknown>
+>
+
+export interface ApiClients {
+    shopperBaskets: ShopperBasketsInstance
```
```suggestion
export interface ApiClients {
    shopperBaskets: ShopperBaskets<{
      clientId: string;
      organizationId: string;
      shortCode: string;
      siteId: string;
    }>
```
If this type is just for what is returned by the provider, then we can include the parameters that are actually set, so that we have the most accurate types when invoking the methods.

vmarta replied:
@kevinxh what do you think of the above comment?

kevinxh replied:
Will's comment makes sense to me. 

the `ShopperBasketsInstance` type is pretty vague, we can make it stricter.

vmarta committed:
Merge branch 'develop' into commerce-api-provider

vmarta committed:
Fix eslint error

vmarta committed:
Document with tsdoc

alexvuong commented on packages/template-typescript-minimal/app/components/_app-config/index.tsx:
```diff
@@ -0,0 +1,35 @@
+            organizationId="f_ecom_zzrf_001"
+            shortCode="8o7m175y"
+            siteId="RefArchGlobal"
+            locale="en_US"
```
Can locale and currency be optional in any case? 🤔

vmarta replied:
Hmm, currently it may look like `locale` and `currency` are not used in the implementation. But I think similar to this line in the retail-react-app, we'll want to require them. A site should have at least 1 locale (and it should have a corresponding currency).

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/5e0cfec12f18bf20cd67166b8727c26ee36b5a0a/packages/template-retail-react-app/app/components/_app-config/index.jsx#L68

cc @kevinxh

kevinxh replied:
for those API endpoints that takes `locale` as arguments, is it optional to the API? I think we want to follow the API specs.

I skimmed through the RAML specs from the isomorphic repo, and it looks like locale is always optional for the endpoints.

wjhsf replied:
Presumably, when a locale/currency is not provided, a default locale/currency is used. And it feels to me like "the locale/currency is <default>, but I want <selected>" is much more common than "the locale/currency is <selected>, but I want <default>". Therefore, to minimize the number of scenarios in which we'd need to explicitly pass the locale/currency, I think it makes sense to have it set in the config.

vmarta committed:
Write a rough test for `useCommerceApi` and the Provider

vmarta committed:
Clean up test

vmarta committed:
Rename to tsx extension

vmarta committed:
Mark the Context as internal use only

The end users would need to call `useCommerceApi` to access the value inside the context.

vmarta commented on packages/commerce-sdk-react/package.json:
```diff
@@ -38,6 +38,8 @@
   "private": true,
   "devDependencies": {
+    "@testing-library/jest-dom": "^5.16.4",
+    "@testing-library/react": "^12.1.5",
```
FYI this version of `testing-library` can't be the latest 13, because it doesn't support our version of React in use. So I downgraded to 12.

kevinxh commented on packages/commerce-sdk-react/src/provider.test.tsx:
```diff
@@ -0,0 +1,37 @@
+        </CommerceApiProvider>
+    )
+
+    expect(screen.getByText(/success/i)).toBeInTheDocument()
```
nit, I'm wondering if we want to assert more than the existence of `shopperSearch`, like we need to make sure the configuration values are properly passed down to those api clients.

vmarta replied:
> like we need to make sure the configuration values are properly passed down to those api clients

Yeah, that can be a test we write later, once we implement the proper initialization of those api clients.

vmarta replied:
Actually, we can write the test now. It should still be valid for later. I've just pushed my commit.

kevinxh commented on packages/commerce-sdk-react/src/provider.tsx:
```diff
@@ -17,8 +32,52 @@ export interface CommerceApiProviderProps {
+            shortCode,
+            siteId
+        },
+        throwOnBadResponse: true
```
🤔 for `throwOnBadResponse`, is there any use case where you don't want it to be true?

wjhsf replied:
`throwOnBadResponse` is the recommended behavior, and is only a flag because we forgot to implement before our first release. 😑 If there's a use case that needs to access the response object even on a bad response, it's available as a property on the error that is thrown.

kevinxh commented on packages/commerce-sdk-react/src/hooks/useCommerceApi.ts:
```diff
@@ -5,48 +5,17 @@
-}
-
+/**
+ * Access a set of Commerce API clients
```
```suggestion
 * Access a set of initialized Commerce API clients, the clients are initialized using the configurations passed to the CommerceApiProvider.
```

I think the key point here is that, the hook returns the **initialized** api clients based on the config you pass to the provider.

kevinxh commented on packages/commerce-sdk-react/src/hooks/useCommerceApi.ts:
```diff
@@ -5,48 +5,17 @@
+/**
+ * Access a set of Commerce API clients
+ *
+ * @returns Commerce API clients
```
I wonder what does it look like on TypeDoc site, do we need to be a bit more specific?

Can we somehow reference the `ApiClients` type? Or is this already the behavior?

cc @alexvuong

wjhsf replied:
Typedoc extracts all of the type information and automatically includes it in the generated docs, so the description doesn't need to include that information.

vmarta committed:
Install types for the correct Node version

vmarta committed:
Clearer comment

vmarta committed:
Stricter type definition for ApiClients

vmarta committed:
Merge branch 'develop' into commerce-api-provider

vmarta committed:
Another test

wjhsf commented on packages/commerce-sdk-react/src/hooks/types.ts:
```diff
@@ -54,16 +54,66 @@ export type ShopperSearchInstance = ShopperSearch<
-    shopperProducts: ShopperProductsInstance
-    shopperPromotions: ShopperPromotionsInstance
-    shopperSearch: ShopperSearchInstance
+    shopperBaskets: ShopperBaskets<{
+        clientId: string
+        organizationId: string
+        shortCode: string
+        siteId: string
+    }>
```
```suggestion
    shopperBaskets: ShopperBaskets<CommerceApiProviderParams>
```
Minor DRY. Could possibly be derived from `CommerceApiProviderProps`, but might take some rekerjiggering to avoid circular deps.

vmarta replied:
Sorry, I should've been more DRY. I've refactored the types in the latest commit.

vmarta committed:
Refactor types

vmarta committed:
Revert changes to typescript-minimal template

vmarta committed:
Move some types back to original place

To mimic the changes coming from #674
